File: swift-bridge/Sources/SpriteKitBridge/Warp.swift
import SpriteKit

private func skWarpPositions(from rawPositions: UnsafePointer<Float>?, count: Int) -> [vector_float2] {
    guard let rawPositions else { return [] }
    var positions: [vector_float2] = []
    positions.reserveCapacity(count)
    for index in 0..<count {
        let base = index * 2
        positions.append(vector_float2(rawPositions[base], rawPositions[base + 1]))
    }
    return positions
}

private func skBorrowWarpable(_ handle: UnsafeMutableRawPointer?) -> (AnyObject & SKWarpable)? {
    guard let handle else { return nil }
    return Unmanaged<AnyObject>.fromOpaque(handle).takeUnretainedValue() as? (AnyObject & SKWarpable)
}

@_cdecl("sk_warp_geometry_grid")
public func sk_warp_geometry_grid() -> UnsafeMutableRawPointer? {
    skRetain(SKWarpGeometryGrid())
}

@_cdecl("sk_warp_geometry_grid_with_dimensions")
public func sk_warp_geometry_grid_with_dimensions(_ columns: Int, _ rows: Int) -> UnsafeMutableRawPointer? {
    skRetain(SKWarpGeometryGrid(__columns: columns, rows: rows, sourcePositions: nil, destPositions: nil))
}

@_cdecl("sk_warp_geometry_grid_with_positions")
public func sk_warp_geometry_grid_with_positions(
    _ columns: Int,
    _ rows: Int,
    _ sourcePositions: UnsafePointer<Float>?,
    _ destPositions: UnsafePointer<Float>?,
    _ count: Int
) -> UnsafeMutableRawPointer? {
    skRetain(SKWarpGeometryGrid(
        __columns: columns,
        rows: rows,
        sourcePositions: skWarpPositions(from: sourcePositions, count: count),
        destPositions: skWarpPositions(from: destPositions, count: count)
    ))
}

@_cdecl("sk_warp_geometry_grid_get_number_of_columns")
public func sk_warp_geometry_grid_get_number_of_columns(_ gridHandle: UnsafeMutableRawPointer?) -> Int {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return grid.numberOfColumns
}

@_cdecl("sk_warp_geometry_grid_get_number_of_rows")
public func sk_warp_geometry_grid_get_number_of_rows(_ gridHandle: UnsafeMutableRawPointer?) -> Int {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return grid.numberOfRows
}

@_cdecl("sk_warp_geometry_grid_get_vertex_count")
public func sk_warp_geometry_grid_get_vertex_count(_ gridHandle: UnsafeMutableRawPointer?) -> Int {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return grid.vertexCount
}

@_cdecl("sk_warp_geometry_grid_get_source_position_x")
public func sk_warp_geometry_grid_get_source_position_x(_ gridHandle: UnsafeMutableRawPointer?, _ index: Int) -> Double {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return Double(grid.sourcePosition(at: index).x)
}

@_cdecl("sk_warp_geometry_grid_get_source_position_y")
public func sk_warp_geometry_grid_get_source_position_y(_ gridHandle: UnsafeMutableRawPointer?, _ index: Int) -> Double {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return Double(grid.sourcePosition(at: index).y)
}

@_cdecl("sk_warp_geometry_grid_get_dest_position_x")
public func sk_warp_geometry_grid_get_dest_position_x(_ gridHandle: UnsafeMutableRawPointer?, _ index: Int) -> Double {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return Double(grid.destPosition(at: index).x)
}

@_cdecl("sk_warp_geometry_grid_get_dest_position_y")
public func sk_warp_geometry_grid_get_dest_position_y(_ gridHandle: UnsafeMutableRawPointer?, _ index: Int) -> Double {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return 0 }
    return Double(grid.destPosition(at: index).y)
}

@_cdecl("sk_warp_geometry_grid_replacing_source_positions")
public func sk_warp_geometry_grid_replacing_source_positions(_ gridHandle: UnsafeMutableRawPointer?, _ sourcePositions: UnsafePointer<Float>?, _ count: Int) -> UnsafeMutableRawPointer? {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return nil }
    return skRetain(grid.replacingBySourcePositions(positions: skWarpPositions(from: sourcePositions, count: count)))
}

@_cdecl("sk_warp_geometry_grid_replacing_dest_positions")
public func sk_warp_geometry_grid_replacing_dest_positions(_ gridHandle: UnsafeMutableRawPointer?, _ destPositions: UnsafePointer<Float>?, _ count: Int) -> UnsafeMutableRawPointer? {
    guard let grid: SKWarpGeometryGrid = skBorrow(gridHandle) else { return nil }
    return skRetain(grid.replacingByDestinationPositions(positions: skWarpPositions(from: destPositions, count: count)))
}

@_cdecl("sk_warpable_get_warp_geometry")
public func sk_warpable_get_warp_geometry(_ nodeHandle: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let node = skBorrowWarpable(nodeHandle),
          let warpGeometry = node.warpGeometry
    else { return nil }
    return skRetain(warpGeometry)
}

@_cdecl("sk_warpable_set_warp_geometry")
public func sk_warpable_set_warp_geometry(_ nodeHandle: UnsafeMutableRawPointer?, _ warpGeometryHandle: UnsafeMutableRawPointer?) {
    guard let node = skBorrowWarpable(nodeHandle) else { return }
    let warpGeometry: SKWarpGeometry? = skBorrow(warpGeometryHandle)
    node.warpGeometry = warpGeometry
}

@_cdecl("sk_warpable_get_subdivision_levels")
public func sk_warpable_get_subdivision_levels(_ nodeHandle: UnsafeMutableRawPointer?) -> Int {
    guard let node = skBorrowWarpable(nodeHandle) else { return 0 }
    return node.subdivisionLevels
}

@_cdecl("sk_warpable_set_subdivision_levels")
public func sk_warpable_set_subdivision_levels(_ nodeHandle: UnsafeMutableRawPointer?, _ levels: Int) {
    guard let node = skBorrowWarpable(nodeHandle) else { return }
    node.subdivisionLevels = levels
}

File: src/warp.rs
use apple_cf::cg::CGPoint;

use crate::effect_node::EffectNode;
use crate::ffi;
use crate::private::handle_type;
use crate::sprite_node::SpriteNode;

handle_type!(WarpGeometry);
handle_type!(WarpGeometryGrid);

pub trait AsWarpGeometry {
    #[doc(hidden)]
    fn as_warp_geometry_ptr(&self) -> *mut core::ffi::c_void;
}

impl AsWarpGeometry for WarpGeometry {
    fn as_warp_geometry_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl AsWarpGeometry for WarpGeometryGrid {
    fn as_warp_geometry_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

#[allow(clippy::cast_possible_truncation)]
fn pack_positions(points: &[CGPoint]) -> Vec<f32> {
    let mut packed = Vec::with_capacity(points.len() * 2);
    for point in points {
        packed.push(point.x as f32);
        packed.push(point.y as f32);
    }
    packed
}

impl WarpGeometryGrid {
    #[must_use]
    pub fn grid() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_warp_geometry_grid()) }
    }

    #[must_use]
    pub fn with_dimensions(columns: usize, rows: usize) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_warp_geometry_grid_with_dimensions(columns, rows)) }
    }

    #[must_use]
    pub fn with_positions(columns: usize, rows: usize, source_positions: &[CGPoint], dest_positions: &[CGPoint]) -> Option<Self> {
        let expected = (columns + 1) * (rows + 1);
        if source_positions.len() != expected || dest_positions.len() != expected {
            return None;
        }
        let source = pack_positions(source_positions);
        let dest = pack_positions(dest_positions);
        unsafe {
            Self::from_raw(ffi::sk_warp_geometry_grid_with_positions(
                columns,
                rows,
                source.as_ptr(),
                dest.as_ptr(),
                expected,
            ))
        }
    }

    #[must_use]
    pub fn number_of_columns(&self) -> usize {
        unsafe { ffi::sk_warp_geometry_grid_get_number_of_columns(self.ptr) }
    }

    #[must_use]
    pub fn number_of_rows(&self) -> usize {
        unsafe { ffi::sk_warp_geometry_grid_get_number_of_rows(self.ptr) }
    }

    #[must_use]
    pub fn vertex_count(&self) -> usize {
        unsafe { ffi::sk_warp_geometry_grid_get_vertex_count(self.ptr) }
    }

    #[must_use]
    pub fn source_position_at(&self, index: usize) -> CGPoint {
        CGPoint::new(
            unsafe { ffi::sk_warp_geometry_grid_get_source_position_x(self.ptr, index) },
            unsafe { ffi::sk_warp_geometry_grid_get_source_position_y(self.ptr, index) },
        )
    }

    #[must_use]
    pub fn dest_position_at(&self, index: usize) -> CGPoint {
        CGPoint::new(
            unsafe { ffi::sk_warp_geometry_grid_get_dest_position_x(self.ptr, index) },
            unsafe { ffi::sk_warp_geometry_grid_get_dest_position_y(self.ptr, index) },
        )
    }

    #[must_use]
    pub fn replacing_source_positions(&self, source_positions: &[CGPoint]) -> Option<Self> {
        if source_positions.len() != self.vertex_count() {
            return None;
        }
        let source = pack_positions(source_positions);
        unsafe { Self::from_raw(ffi::sk_warp_geometry_grid_replacing_source_positions(self.ptr, source.as_ptr(), source_positions.len())) }
    }

    #[must_use]
    pub fn replacing_dest_positions(&self, dest_positions: &[CGPoint]) -> Option<Self> {
        if dest_positions.len() != self.vertex_count() {
            return None;
        }
        let dest = pack_positions(dest_positions);
        unsafe { Self::from_raw(ffi::sk_warp_geometry_grid_replacing_dest_positions(self.ptr, dest.as_ptr(), dest_positions.len())) }
    }
}

pub trait WarpableNode {
    #[doc(hidden)]
    fn as_warpable_ptr(&self) -> *mut core::ffi::c_void;

    #[must_use]
    fn warp_geometry(&self) -> Option<WarpGeometry> {
        unsafe { WarpGeometry::from_raw(ffi::sk_warpable_get_warp_geometry(self.as_warpable_ptr())) }
    }

    fn set_warp_geometry<W: AsWarpGeometry>(&self, warp: Option<&W>) {
        unsafe {
            ffi::sk_warpable_set_warp_geometry(
                self.as_warpable_ptr(),
                warp.map_or(core::ptr::null_mut(), AsWarpGeometry::as_warp_geometry_ptr),
            );
        };
    }

    #[must_use]
    fn subdivision_levels(&self) -> isize {
        unsafe { ffi::sk_warpable_get_subdivision_levels(self.as_warpable_ptr()) }
    }

    fn set_subdivision_levels(&self, levels: isize) {
        unsafe { ffi::sk_warpable_set_subdivision_levels(self.as_warpable_ptr(), levels) };
    }
}

impl WarpableNode for SpriteNode {
    fn as_warpable_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl WarpableNode for EffectNode {
    fn as_warpable_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}
